app/src/nxt/java/com/pileproject/drive/execution/NxtCarController.java

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (C) 2011-2016 PILE Project, Inc. <dev@pileproject.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.pileproject.drive.execution;
+
+import com.pileproject.drive.preferences.MachinePreferences;
+import com.pileproject.drive.preferences.MachinePreferencesSchema;
+import com.pileproject.drivecommand.model.nxt.NxtMachine;
+import com.pileproject.drivecommand.model.nxt.port.NxtInputPort;
+import com.pileproject.drivecommand.model.nxt.port.NxtOutputPort;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import static com.pileproject.drive.app.DriveApplication.getContext;
+import static com.pileproject.drive.execution.CarControllerBase.INPUT.LINE;
+import static com.pileproject.drive.execution.CarControllerBase.INPUT.SOUND;
+import static com.pileproject.drive.execution.CarControllerBase.INPUT.TOUCH;
+
+public class NxtCarController extends CarControllerBase {
+    private final boolean mIsLejosFirmware;
+
+    /**
+     * binds each sensor and motor to their port
+     */
+    public NxtCarController(NxtMachine machine) {
```

Comment:
> Would it be better that the class has no dependency on the preference?

So, how do we deal with it? Do you mean that we should connect devices with ports by some setter methods? I'm sorry I may not understand.


> And should NxtMachine know firmware information? 

Hmm, I'm worried about it may introduce some complexities to the library [drivecommand] because leJOS is just a variant of firmware, though I'm not so familiar wth others. Supporting only the standard firmware is not bad I think.